File: 프로그래머스/2/155651. 호텔 대실/호텔 대실.py
def solution(book_time):
    answer = 0
    book_time.sort()
    room = []
    
    # 이렇게 하면 10분 더했을 때 60분이 넘어가면 시를 올려줘야하는데 문제발생
    # for book in book_time:
    #     start = int(book[0].replace(':', ''))
    #     end = int(book[1].replace(':', '')) + 10  # 청소시간까지
    
    # 다 분으로 바꿔
    for book in book_time:
        start_h, start_m = book[0].split(':')
        start = int(start_h) * 60 + int(start_m)
        end_h, end_m = book[1].split(':')
        end = int(end_h) * 60 + int(end_m) + 10
        
        possible = False
        for i in range(len(room)):
            if room[i] <= start:
                room[i] = end
                possible = True  # 들어갈 수 있는 방 탐색
                break
        if not possible:
            room.append(end)  # 방 추가
    
    return len(room)

File: 프로그래머스/2/155651. 호텔 대실/호텔 대실.java
import java.util.*;

class Solution {
    public int solution(String[][] book_time) {
        List<int[]> times = new ArrayList<>();
        PriorityQueue<Integer> room = new PriorityQueue<>();
        for (String[] bt : book_time) {
            String startTime = bt[0];
            String endTime = bt[1];
            times.add(new int[]{timeToMinute(startTime), timeToMinute(endTime) + 10});
        }
        times.sort((a, b) -> a[0] - b[0]);
        
        int answer = 0;
        for (int[] time : times) {
            int start = time[0];
            int end = time[1];
            if (room.isEmpty()) {
                answer++;
                room.add(end);
                continue;
            }
            int t = room.peek();
            if (t <= start)
                room.poll();
            else
                answer++;
            room.add(end);
        }
        
        return answer;
    }
    
    private int timeToMinute(String time) {
        String[] times = time.split(":");
        int hour = Integer.parseInt(times[0]) * 60;
        int minute = Integer.parseInt(times[1]);
        return hour + minute;
    }
}